clang/lib/Driver/ToolChains/AMDGPUOpenMP.cpp

```
@@ -68,7 +68,9 @@ llvm::opt::DerivedArgList *AMDGPUOpenMPToolChain::TranslateArgs(
     Action::OffloadKind DeviceOffloadKind) const {
   DerivedArgList *DAL =
       HostTC.TranslateArgs(Args, BoundArch, DeviceOffloadKind);
-  if (!DAL || Args.hasArg(options::OPT_fsanitize_EQ))
+  // Skip sanitize options passed from the HostTC. The decision to instrument
+  // device code is computed only by 'shouldSkipSanitizeOption'.
+  if (!DAL && DAL->hasArg(options::OPT_fsanitize_EQ))
```

I don't fully understand why we need this. Shouldn't the `shouldSkip` part handle this just fine?

---

Sorry, actually here the DAL computed from the HostTC picks up the sanitize option like ```-fsanitize=address``` and pass forward to ```DAL``` of device toolchain. The ```shouldSkipSanitizeOption ```  runs over each argument of ```Args [The user provided arguments which are not derived]``` and checks if the ```-fsanitize=address``` is there it early returns false which then avoids checking the :xnack+ capability. 

So,  assume the  case where when we pass ```-fsanitize=address --offload-arch=gfx90a``` . This should avoid appending  ```-fsanitize=address``` in the DAL for device toolchain but it dosen't since the DAL already has -fsanitize=address from host toolchain and due to early return from ```shouldSkipSanitizeOption``` where it checks for the ```:xnack+``` capability is skipped.

---

What I have seen for HIPAMDToolChain , the DAL is always constructed every time dynamically allocated every time.

---

Should The ```shouldSkipSanitizeOption``` consider an extra parameter ```DAL``` allocated by the host and resolve that issue ? @jhuber6  and @yxsamliu  is that the correct way ,anything I am missing @yxsamliu ?

---

> Sorry I'm still confused. If the user cannot enable `-fsanitize=address` if the GPU target doesn't support `xnack` then that should be separate driver check / error. Unless we're just silently ignoring that?

Yes we are silently ignoring that and that ```-fsanitize=address``` is there for the host only, but for device it is skipped.

---

Silently ignoring this sounds pretty awful to me, but if that's what ROCm wants I guess that's the behavior.

---

If the goal is to ignore it, why can't we just instead be like `if (TCArgs.hasFlag(...) && isValid)` on every use instead of hacking around it in the argument list.

---

We actually in check it in this part of code https://github.com/llvm/llvm-project/blob/752ebec054f8b0bf7c221813724984f10096355a/clang/lib/Driver/ToolChains/AMDGPU.cpp#L1118

---

I just don't understand the rationale for removing it from the argument list instead of just guarding wherever we look up this flag.

---

> if it's not legal for some case, then emit an error. If the user doesn't want it then they can do `-Xarch-host -fsanitize=address`.

We actually emit the warning not error, since it's legal for the host to be asan instrumented but for device it depends based on ```--offload-arch=``` containing :xnack+. If :xnack+ then allow it otherwise emit warning silently skip it.

---

So, is it prohibitively difficult to do something more akin to.

```c
if (Args.hasFlag(OPT_fsanitize_address, OPT_fno_sanitize, false) &&
    Args.hasFlag(OPT_fgpu_sanitize, OPT_fno_gpu_sanitize, true)) {
  if (isSanitizerLegal(Args))
    // something
  else
    Diags.warn("incompatible, ignoring");    
}
```